src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs

```diff
@@ -204,6 +204,28 @@ public void Node_Table_Doesnt_Modify_Single_Item_Multiple_Times_When_Same()
             AssertTableEntries(newTable, expected);
         }
 
+        [Fact]
+        public void Node_Table_Caches_Previous_Object_When_Modification_Considered_Cached()
+        {
+            var builder = NodeStateTable<int>.Empty.ToBuilder();
+            builder.AddEntries(ImmutableArray.Create(1), EntryState.Added);
+            builder.AddEntries(ImmutableArray.Create(2), EntryState.Added);
+            builder.AddEntries(ImmutableArray.Create(3), EntryState.Added);
+            var previousTable = builder.ToImmutableAndFree();
+
+            var expected = ImmutableArray.Create((1, EntryState.Added), (2, EntryState.Added), (3, EntryState.Added));
+            AssertTableEntries(previousTable, expected);
+
+            builder = previousTable.ToBuilder();
+            Assert.True(builder.TryModifyEntry(1, EqualityComparer<int>.Default));           // ((1, EntryState.Cached))
```

Comment:
So I keep forgetting: is the entry we're trying to modify denoted implicitly by the order of calls to `TryModifyEntry`?